### PR TITLE
feat(vulnerability): assemble vulnerabilities info for artifact

### PR DESCRIPTION
**Pull request by heww**
1. Assemble scan overview to artifact when scanner enabled in the
project of the artifact.
2. Set addition link for vulnerabilities to artifact when scanner
enabled in the project of the artifact.

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/api/artifact/controller.go [hunk 1]
```diff
@@ -516,6 +521,16 @@ func (c *controller) assembleArtifact(ctx context.Context, art *artifact.Artifac
 	artifact := &Artifact{
 		Artifact: *art,
 	}
+
+	if artifact.RepositoryName == "" {
```
Comment (wy65701436) [inline src/api/artifact/controller.go]: I'd prefer to store the repository name into DB.

File: src/server/v2.0/handler/artifact.go [hunk 2]
```diff
@@ -243,14 +253,23 @@ func (a *artifactAPI) GetAddition(ctx context.Context, params operation.GetAddit
...
 		return a.SendError(ctx, err)
 	}
+
 	artifact, err := a.artCtl.GetByReference(ctx, fmt.Sprintf("%s/%s", params.ProjectName, params.RepositoryName), params.Reference, nil)
 	if err != nil {
 		return a.SendError(ctx, err)
 	}
-	addition, err := a.artCtl.GetAddition(ctx, artifact.ID, strings.ToUpper(params.Addition))
+
+	var addition *resolver.Addition
```
Comment (wy65701436) [inline src/server/v2.0/handler/artifact.go]: it seems that it breaks the object abstrator layer? @ywk253100 your thoughts?

File: src/server/v2.0/handler/assembler/vul.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	withScanOverview bool
+}
+
+func (assembler *VulAssembler) hasScanner(ctx context.Context, projectID int64) bool {
+	value, ok := assembler.scanners[projectID]
+	if !ok {
+		scanner, err := assembler.scannerCtl.GetRegistrationByProject(projectID)
+		if err != nil {
+			return false
```
Comment (wy65701436) [inline src/server/v2.0/handler/assembler/vul.go]: log err here？

File: src/server/v2.0/handler/assembler/vul.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// Assemble assemble vul for the artifacts
+func (assembler *VulAssembler) Assemble(ctx context.Context) error {
+	version := internal.GetAPIVersion(ctx)
+
+	for _, artifact := range assembler.artifacts {
+		hasScanner := assembler.hasScanner(ctx, artifact.ProjectID)
+
+		if hasScanner {
```
Comment (wy65701436) [inline src/server/v2.0/handler/assembler/vul.go]: what about to refactor it as

```
if !hasScanner{
    continue
}
if assembler.withScanOverview {
 ....
}
```

File: src/api/artifact/controller.go (hunk 1 above)
Comment (ywk253100) [reply src/api/artifact/controller.go]: Agree. I opened an issue to track the work https://github.com/goharbor/harbor/issues/10805

File: src/api/artifact/controller.go [hunk 5]
```diff
@@ -516,6 +521,16 @@ func (c *controller) assembleArtifact(ctx context.Context, art *artifact.Artifac
...
+		repo, err := c.repoMgr.Get(ctx, artifact.RepositoryID)
+		if err != nil {
+			log.Errorf("repository %d not found", artifact.RepositoryID)
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: The error message is confused as it isn't alway the not found error

File: src/server/v2.0/handler/util.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+				return nil, err
+			}
+
+			vulnerabilities[rp.MimeType] = vrp
+		}
+	}
+
+	// TODO: handle errors for scanner not found in the project and report not found for the artifact
+
+	content, _ := json.Marshal(vulnerabilities)
```
Comment (wy65701436) [inline src/server/v2.0/handler/util.go]: why do not handle the err?

File: src/server/v2.0/handler/assembler/vul.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+		if hasScanner && assembler.withScanOverview {
+			art := &v1.Artifact{
+				NamespaceID: artifact.ProjectID,
+				Repository:  artifact.RepositoryName,
+				Digest:      artifact.Digest,
+				MimeType:    artifact.ManifestMediaType,
+			}
+
+			overview, err := assembler.scanCtl.GetSummary(art, []string{v1.MimeTypeNativeReport})
+			if err == nil && len(overview) > 0 {
```
Comment (wy65701436) [inline src/server/v2.0/handler/assembler/vul.go]: log err?

**Pull request merged**